File: csharp/solution/Webshop.Service/Payment/PaymentService.cs
using Webshop.Service.Common;
using Webshop.Service.Users;

namespace Webshop.Service.Payment;

public interface IPaymentService
{
    PaymentResult HandlePayment(string userId, decimal amountInNok);
}

public class PaymentService : IPaymentService
{
    private readonly IThirdPartyPaymentIntegration _thirdPartyPaymentIntegration;
    private readonly IUserService _userService;
    private readonly INotificationService _notificationService;
    private readonly ILogger? _logger;

    public PaymentService(IThirdPartyPaymentIntegration thirdPartyPaymentIntegration, IUserService userService, INotificationService notificationService, ILogger? logger)
    {
        _thirdPartyPaymentIntegration = thirdPartyPaymentIntegration;
        _userService = userService;
        _notificationService = notificationService;
        _logger = logger;
    }

    public PaymentResult HandlePayment(string userId, decimal amountInNok)
    {
        var user = _userService.GetUser(userId);
        if (user == null)
        {
            var msg = $"Attempted to handle payment for non-existing user {userId}";
            _logger?.LogError(msg);
            return new PaymentResult { Code = PaymentResultCode.UserNotFound, ErrorMessage = msg };
        }

        if (amountInNok < 0)
        {
            return new PaymentResult { Code = PaymentResultCode.InvalidAmount, ErrorMessage = "Amount cannot be negative" };
        }

        var paymentOk = _thirdPartyPaymentIntegration.HandlePaymentForUser(userId, amountInNok, "WebShop payment");
        if (!paymentOk)
        {
            var msg = $"Payment failed for user {userId} with amount {amountInNok}";
            _logger?.LogError(msg);
            return new PaymentResult { Code = PaymentResultCode.PaymentIntegrationFailed, ErrorMessage = msg };
        }

        _notificationService.SendMail(user.Email, "Payment succeeded");

        return new PaymentResult { Code = PaymentResultCode.Success };
    }
}

public class PaymentResult
{
    public bool Success => Code == PaymentResultCode.Success;
    public PaymentResultCode Code { get; set; }
    public string? ErrorMessage { get; set; }
}

public enum PaymentResultCode
{
    Success,
    UserNotFound,
    InvalidAmount,
    PaymentIntegrationFailed
}

File: java/exercises/src/main/java/Webshop/Service/Payment/PaymentService.java
package Webshop.Service.Payment;

import Webshop.Service.Common.Logger;
import Webshop.Service.Common.NotificationService;
import Webshop.Service.Payment.PaymentResult;
import Webshop.Service.Payment.PaymentResultCode.*;
import Webshop.Service.Users.UserService;
import Webshop.Service.Users.User;

//import static Webshop.Service.Common.Logger.logError;

interface IPaymentService {
    PaymentResult handlePayment(String userId, double amountInNok);
}

public class PaymentService implements IPaymentService {
    private final ThirdPartyPaymentIntegration thirdPartyPaymentIntegration;
    private final UserService userService;
    private final NotificationService notificationService;
    private final Logger logger;

    public PaymentService(ThirdPartyPaymentIntegration thirdPartyPaymentIntegration, UserService userService, NotificationService notificationService, Logger logger) {
        this.thirdPartyPaymentIntegration = thirdPartyPaymentIntegration;
        this.userService = userService;
        this.notificationService = notificationService;
        this.logger = logger;
    }

    public PaymentResult handlePayment(String userId, double amountInNok) {
        User user = userService.getUser(userId);
        if (user == null) {
            String msg = String.format("Attempted to handle payment for non-existing user %s", userId);
            logger.logError(msg);
            return new PaymentResult(PaymentResultCode.USER_NOT_FOUND, msg);
        }

        if (amountInNok < 0) {
            return new PaymentResult(PaymentResultCode.INVALID_AMOUNT, "Amount cannot be negative");
        }

        boolean paymentOk = thirdPartyPaymentIntegration.handlePaymentForUser(userId, amountInNok, "WebShop payment");
        if (!paymentOk) {
            String msg = String.format("Payment failed for user %s with amount %.2f", userId, amountInNok);
            logger.logError(msg);
            return new PaymentResult(PaymentResultCode.PAYMENT_INTEGRATION_FAILED, msg);
        }

        notificationService.sendMail(user.getEmail(), "Payment succeeded");

        return new PaymentResult(PaymentResultCode.SUCCESS);
    }
}

